sys/fido2/ctap/ctap_mem.c

```diff
@@ -154,3 +161,13 @@ int fido2_ctap_mem_get_offset_of_rk_into_flashpage(uint16_t rk_idx)
 
     return -1;
 }
+
+unsigned fido2_ctap_mem_get_state_flashpage_number(void)
+{
+    return flashpage_last_free() - CONFIG_FIDO2_CTAP_NUM_FLASHPAGES + 1;
```

Comment:
Are we sure that this will be below `flashpage_first_free()`?

Reply:
No. Should I add a check into [ctap_mem_init](https://github.com/RIOT-OS/RIOT/blob/5e925e1906a7d10b2ab6f8d9b972755b8d290d19/sys/fido2/ctap/ctap_mem.c#L48) checking this condition and returning an error if it doesn't hold ?

Reply:
Side question: Why do follow the logic of writing to the end of the flash at all? Why not use first free?

Reply:
Because `flashpage_first_free()` will return a different value across firmware updates if riotboot is used in two-slot mode. This could result in the loss of FIDO2 data.

Reply:
But `flashpage_last_free()` also returns a different value across firmware updates (always the end of the current slot)

Reply:
Oh really ? Then I misunderstood how the riotboot slots work. From the [documentation](https://api.riot-os.org/group__bootloader__riotboot.html) I would have thought that slot 0 and slot 1 are at the beginning of the flash after the bootloader and the rest of the flash is unused. In this case `flashpage_last_free()` would always be the same.

Reply:
With riotboot the flash is split in two halves, the first half is slot 0 and the second half is slot1.

You can't have the slots have dynamic sizes as otherwise a firmware update would not be allowed to be different in size than the previous version.

btw: You added [a warning](https://github.com/RIOT-OS/RIOT/blob/master/drivers/include/periph/flashpage.h#L225) to `flashpage_last_free()` yourself :wink:

Reply:
Thanks for the clarification !

Reply:
So should I change it to use `flashpage_first_free()`? And what about the case where `flashpage_first_free()` is not above `flashpage_first_free()` ?

Reply:
If I understand it correctly, neither `flashpage_first_free` nor `flashpage_last_free` works in the case of firmware updates. Is that correct?
- Can we reserve a flash region prior to assigning slot 0/1? There are other use cases with such requirements (#17092).
- Side note: FIDO2 keys that are currently stored in flash will likely go to a secure element in future (WIP). So the remaining data to be stored in flash is only configurations, I suppose.
- For the FIDO2 case: Do we actually need to consider firmware updates? How would a deployment look like? Otherwise, we can ignore this problem and simply document it properly.

Reply:
Fixed the code to use `flashpage_first_free()`.
I agree with @PeterKietzmann that firmware updates should not be considered as of now.